server/channels/app/post.go

```diff
@@ -773,51 +773,65 @@ func (a *App) publishWebsocketEventForPermalinkPost(c request.CTX, post *model.P
 	}
 
-	permalinkPreviewedChannel, err := a.GetChannel(c, previewedPost.ChannelId)
-	if err != nil {
-		if err.StatusCode == http.StatusNotFound {
+	permalinkPreviewedChannel, appErr := a.GetChannel(c, previewedPost.ChannelId)
+	if appErr != nil {
+		if appErr.StatusCode == http.StatusNotFound {
 			mlog.Warn("channel containing permalinked post not found", mlog.String("referenced_channel_id", previewedPost.ChannelId))
 			return false, nil
 		}
-		return false, err
+		return false, appErr
 	}
 
 	permalinkPreviewedPost := post.GetPreviewPost()
-	for _, cm := range channelMembers {
-		if permalinkPreviewedPost != nil {
-			post.Metadata.Embeds[0].Data = permalinkPreviewedPost
-		}
+	if permalinkPreviewedPost != nil {
+		post.Metadata.Embeds[0].Data = permalinkPreviewedPost
+	} else {
+		return false, nil
+	}
 
-		postForUser := a.sanitizePostMetadataForUserAndChannel(c, post, permalinkPreviewedPost, permalinkPreviewedChannel, cm.UserId)
+	// This is a hack to cache the JSON representation of the post metadata
+	// to be attached to the websocket event if the user doesn't have permission.
+	// We wipe the data, get the JSON, and restore the data again.
+	// This simplifies the data handling in the channel hook, where otherwise
+	// multiple goroutines would be racing to modify the Metadata field.
+	// And deep-copying the Embed data also has challenges because the type is "any".
+	post.Metadata.Embeds[0].Data = nil
+	modifiedJSON, err := post.ToJSON()
+	if err != nil {
+		mlog.Warn("Failed to encode post to JSON", mlog.Err(err))
+	}
+	post.Metadata.Embeds[0].Data = permalinkPreviewedPost
 
-		// Using DeepCopy here to avoid a race condition
-		// between publishing the event and setting the "post" data value below.
-		messageCopy := message.DeepCopy()
-		broadcastCopy := messageCopy.GetBroadcast()
-		broadcastCopy.UserId = cm.UserId
-		messageCopy.SetBroadcast(broadcastCopy)
+	// Modifying post for the original sender, in case they don't have permission.
+	// This wipes off the original post content. And then the marshalled post content
+	// in the websocket event is taken care of in the channel hook.
+	if permalinkPreviewedChannel != nil && !a.HasPermissionToReadChannel(c, post.UserId, permalinkPreviewedChannel) {
+		post.Metadata.Embeds[0].Data = nil
+	}
 
-		postJSON, jsonErr := postForUser.ToJSON()
-		if jsonErr != nil {
-			mlog.Warn("Failed to encode post to JSON", mlog.Err(jsonErr))
+	broadcastCopy := message.GetBroadcast()
+	broadcastCopy.ChannelHook = func(userID string, ev *model.WebSocketEvent) bool {
+		hasChange := a.shouldSanitizePostMetadataForUserAndChannel(c, post, permalinkPreviewedChannel, userID)
+		if !hasChange {
+			// If there is no change, then the original post which was attached
+			// will get sent.
+			return false
 		}
-		messageCopy.Add("post", postJSON)
-		a.Publish(messageCopy)
+
+		ev.AddWithCopy("post", modifiedJSON)
+		return true
 	}
+	message.SetBroadcast(broadcastCopy)
+	a.Publish(message)
```

Comment:
I once again spent too long wondering where the `"post"` was attached before remembering it was on the caller's side. </datapoint>